### ISSUE TITLE
fail fo find nodes if schemas are declared

### ISSUE THREAD

**Issue by domingoscisci** [r-lib/xml2#222]
Hi,

### Issue Description and Expected Result
I've found the following issue: If I try to read a XML file/txt which includes a schema declaration, the library fails to find the node I'm looking for.

### Reproducible Example
```
library(xml2)
x<- "<?xml version=\"1.0\" encoding=\"UTF-8\"?><codeBook xmlns=\"ddi:codebook:2_5\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"ddi:codebook:2_5 http://www.ddialliance.org/Specification/DDI-Codebook/2.5/XMLSchema/codebook.xsd\"><docDscr><citation><titlStmt><titl xml:lang=\"it\">Some text</titl></titlStmt></citation></docDscr></codeBook>"
xml<- read_xml(x)
xml_find_all(xml, ".//titl")
```
**_Result:_ {xml_nodeset (0)}**

```
y <- "<?xml version=\"1.0\" encoding=\"UTF-8\"?><codeBook><docDscr><citation><titlStmt><titl xml:lang=\"it\">Some text</titl></titlStmt></citation></docDscr></codeBook>"
xml <- read_xml(y)
xml_find_all(xml, ".//titl")
```
**_Result:_
{xml_nodeset (1)}
[1] <titl xml:lang="it">Some text</titl>**

<details>
<summary>Session Info</summary>

```r
R version 3.5.0 (2018-04-23)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=Italian_Italy.1252  LC_CTYPE=Italian_Italy.1252    LC_MONETARY=Italian_Italy.1252 LC_NUMERIC=C                   LC_TIME=Italian_Italy.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] xml2_1.2.0

loaded via a namespace (and not attached):
[1] compiler_3.5.0 tools_3.5.0    yaml_2.1.19    Rcpp_0.12.16
```
</details>

Thanks for your help.
Domingo

**Comment by cderv** [comment]
hI think te behavior you are observing is by design, as explained in some examples in documentation.  
xml can have namespace and xml2 :package: can handle them but you need to specifiy the namespace prefix to get the correct node.
See documenation for xml2 :package: about [XML namespaces](http://xml2.r-lib.org/reference/xml_ns.html) and [`xml_find_all()`](http://xml2.r-lib.org/reference/xml_find_all.html)
Here is some examples on how to deal with namespace

``` r
library(xml2)
x<- "<?xml version=\"1.0\" encoding=\"UTF-8\"?><codeBook xmlns=\"ddi:codebook:2_5\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"ddi:codebook:2_5 http://www.ddialliance.org/Specification/DDI-Codebook/2.5/XMLSchema/codebook.xsd\"><docDscr><citation><titlStmt><titl xml:lang=\"it\">Some text</titl></titlStmt></citation></docDscr></codeBook>"
xml<- read_xml(x)
```

Your xml namespace is correctly retrieved with some default name when not named in _xmlns_
``` r
(ns <- xml_ns(xml))
#> d1  <-> ddi:codebook:2_5
#> xsi <-> http://www.w3.org/2001/XMLSchema-instance
```

You can rename namespace to help by having informative name

``` r
(ns_renamed <- xml_ns_rename(xml_ns(xml), d1 = "ddi"))
#> ddi <-> ddi:codebook:2_5
#> xsi <-> http://www.w3.org/2001/XMLSchema-instance
```

namespace is used to qualify nodes

``` r
# get titl node children
title_node <- xml_children(xml_children(xml_children(xml_children(xml))))
# name without namespace
xml_name(title_node)
#> [1] "titl"
# full name with namespace
xml_name(title_node, ns)
#> [1] "d1:titl"
xml_name(title_node, ns_renamed)
#> [1] "ddi:titl"
```

If you do not precise namespace, you cannot retrieve node <titl> associated with namespace ddi

``` r
# this is not working
xml_find_all(xml, ".//titl")
#> {xml_nodeset (0)}
# this is working
xml_find_all(xml, ".//ddi:titl", ns_renamed)
#> {xml_nodeset (1)}
#> [1] <titl xml:lang="it">Some text</titl>
```

If you don’t want to deal with namespace because, for example, you are sure there is no conflict in names, you can strip it off. No need to precise the namespace prefix then.

``` r
# take off the namespace in the parsed xml
xml_ns_strip(xml)
# This is now working without using namespace prefix
xml_find_all(xml, ".//titl")
#> {xml_nodeset (1)}
#> [1] <titl xml:lang="it">Some text</titl>
```

In conclusion, when present, you need to use namespace in XPATH when using namespace in xml or deals with the namespace before moving into xml tree.

Created on 2018-05-09 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by domingoscisci** [comment]
Dear @cderv,
thank you for your reply. I missed that part of the documentation. Now it's all clear to me.
Thanks!
Domingo

**Comment by cderv** [comment]
Glad I could help! 
There is always room for improvement in documentation, so if you think it needs some, please feel free to suggest. 
Otherwise, if it is ok for you, you may close the issue.
